core/distributed/sparse_communicator.cpp

```diff
@@ -0,0 +1,275 @@
+
+    // adjacent constructor guarantees that querying sources/destinations
+    // will result in the array having the same order as defined here
+    MPI_Comm new_comm;
+    MPI_Info info;
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Info_dup(MPI_INFO_ENV, &info));
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Dist_graph_create_adjacent(
+        base.get(), in_degree, sources_host->get_const_data(), MPI_UNWEIGHTED,
+        out_degree, destinations_host->get_const_data(), MPI_UNWEIGHTED, info,
+        false, &new_comm));
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Info_free(&info));
+
+    return mpi::communicator::create_owning(new_comm, base.force_host_buffer());
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+sparse_communicator::sparse_communicator(
+    mpi::communicator comm,
+    const index_map<LocalIndexType, GlobalIndexType>& imap)
+    : default_comm_(MPI_COMM_NULL),
+      recv_sizes_(imap.get_remote_local_idxs().size()),
+      recv_offsets_(recv_sizes_.size() + 1)
+{
+    auto exec = imap.get_executor();
+
+    auto& recv_target_ids = imap.get_remote_target_ids();
+    std::transform(imap.get_remote_global_idxs().begin(),
+                   imap.get_remote_global_idxs().end(), recv_sizes_.begin(),
+                   [](const auto& a) { return a.get_size(); });
+    auto send_envelope =
+        communicate_inverse_envelope(exec, comm, recv_target_ids, recv_sizes_);
+    auto& send_target_ids = std::get<0>(send_envelope);
+    send_sizes_ = std::move(std::get<1>(send_envelope));
+    send_offsets_.resize(send_sizes_.size() + 1);
+
+    std::partial_sum(recv_sizes_.begin(), recv_sizes_.end(),
+                     recv_offsets_.begin() + 1);
+    std::partial_sum(send_sizes_.begin(), send_sizes_.end(),
+                     send_offsets_.begin() + 1);
+
+    send_idxs_ = communicate_send_gather_idxs(
+        comm, imap.get_remote_local_idxs().get_flat(), recv_target_ids,
+        recv_sizes_, send_target_ids, send_sizes_);
+
+    default_comm_ =
+        create_neighborhood_comm(comm, recv_target_ids, send_target_ids);
+}
+
+#define GKO_DECLARE_SPARSE_COMMUNICATOR(LocalIndexType, GlobalIndexType) \
+    sparse_communicator::sparse_communicator(                            \
+        mpi::communicator comm,                                          \
+        const index_map<LocalIndexType, GlobalIndexType>& imap)
+
+GKO_INSTANTIATE_FOR_EACH_LOCAL_GLOBAL_INDEX_TYPE(
+    GKO_DECLARE_SPARSE_COMMUNICATOR);
+
+#undef GKO_DECLARE_SPARSE_COMMUNICATOR
+
+
+std::shared_ptr<const Executor> get_mpi_exec(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm)
+{
+    bool use_host_buffer = mpi::requires_host_buffer(exec, comm);
+    return use_host_buffer ? exec->get_master() : exec;
+}
+
+
+template <typename ValueType>
+mpi::request sparse_communicator::communicate(
+    const matrix::Dense<ValueType>* local_vector,
+    const detail::DenseCache<ValueType>& send_buffer,
+    const detail::DenseCache<ValueType>& recv_buffer) const
```

Comment:
Looking at this now, I dont get why the sparse_communicator class handles this specific communicate function.